### ISSUE TITLE
Change Request: Publish types

### ISSUE THREAD

**Issue by nzakas** [eslint/eslint#18845]
### ESLint version

HEAD

### What problem do you want to solve?

Right now, we're in a situation where there's a `@types/eslint` package that people use for type definitions for ESLint. This package is very popular (12M downloads per week), but it's always out of date because we don't maintain it. This continues to cause issues as we work on language plugins and the core rewrite.

Further, it can take weeks to get changes to `@types/eslint` approved([last PR](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/69957)) and deployed. Because we'll likely have more type changes as we accelerate language plugins and core rewrite work, this just doesn't seem sustainable.

### What do you think is the correct solution?

ESLint should publish its own types.

I think the easiest place to start is just to copy the type definitions from `@types/eslint` as-is and publish those as part of the `eslint` package. This will be provide a good level of compatibility for those using `@types/eslint`.

We can then easily make modifications, including references to `@eslint/core` as we go forward.

In terms of types from `eslint` and `@eslint/core`:

* `@eslint/core` types will be exclusively those types that are for use in the core rewrite. This excludes API interfaces such as `ESLint` and `Linter`.
* `eslint` types will be a superset of `@eslint/core` that includes API interfaces for the `eslint` package.

### Participation

- [X] I am willing to submit a pull request for this change.

**Comment by JoshuaKGoldberg** [comment]
👍 from me on this. Having separate types in `@types/*` packages is a common source of confusion &/| inconvenience on the TypeScript/types side. The standard DefinitelyTyped recommendation is to bundle types in packages when the package is _"written in TypeScript"[^1]_ - but I think that that phrasing should probably be updated to _"has its own type definitions"_.

Once ESLint has its own types, `@types/eslint` can become a "stub" package. DefinitelyTyped has built-in support for removing packages[^2].

One cautionary note is that changing types layouts -even moving from DT to a first-party plugin- can impact a lot of downstream repos in surprising ways. I'd suggest trying the PR that resolves this issue on a few community ESLint configs & plugins that use `@types/eslint`.

[^1]: https://github.com/DefinitelyTyped/DefinitelyTyped?tab=readme-ov-file#create-a-new-package
[^2]: https://github.com/DefinitelyTyped/DefinitelyTyped?tab=readme-ov-file#removing-a-package

**Comment by nzakas** [comment]
Thanks for the insights @JoshuaKGoldberg, those are very helpful.

And agree, we should definitely test with some downstream packages. Hopefully by starting out with the same type definitions, we'll minimize any disturbance.

**Comment by nzakas** [comment]
A complicating factor:
https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/eslint/rules

These look to be about three years out-of-date, so I wonder if it would be safe to just drop them at this point.

**Comment by ericmorand** [comment]
That would be awesome. I'd like to add that working with DefinitelyTyped is a very unpleasing process because of the tremendous (and growing) size of the repository that slaughters basically every computer I've used to provide some fixes there. I'm not ashamed to say that I stopped maintaining some types there because working on the repository is a pain. So, having the types maintained here would also, in my opinion, make their maintenance way easier and collaboration-friendly.

**Comment by fasttime** [comment]
> A complicating factor: https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/eslint/rules
 
I think part of those types could be autogenerated from rule `meta.schema` arrays, but there will be discrepancies.

**Comment by nzakas** [comment]
@fasttime they could be, but it seems like that would take a lot of time to verify.

To move forward here, there are two paths:

1. Include those files as-is without worrying about whether or not they are up to date
2. Don't include them as just export `ESLintRules` as:

```ts
export interface ESLintRules
    extends
        Linter.RulesRecord
{}
```
